tests/test_expression.py

```diff
@@ -1157,6 +1157,38 @@ def test_variable_clone(initype, inireftype, newtype, newreftype):
     var = var.clone(type=newtype)  # pylint: disable=no-member
     assert isinstance(var, newreftype)
 
+@pytest.mark.parametrize('initype,newtype,reftype', [
+    # Preserve existing type info if type=None is given
+    (SymbolAttributes(BasicType.REAL), None, SymbolAttributes(BasicType.REAL)),
+    (SymbolAttributes(BasicType.INTEGER), None, SymbolAttributes(BasicType.INTEGER)),
+    (SymbolAttributes(BasicType.DEFERRED), None, SymbolAttributes(BasicType.DEFERRED)),
+    (SymbolAttributes(BasicType.DEFERRED, intent='in'), None,
+     SymbolAttributes(BasicType.DEFERRED, intent='in')),
+    # Update from deferred to known type
+    (SymbolAttributes(BasicType.DEFERRED), SymbolAttributes(BasicType.INTEGER),
+     SymbolAttributes(BasicType.INTEGER)),
+    (SymbolAttributes(BasicType.DEFERRED), SymbolAttributes(BasicType.REAL),
+     SymbolAttributes(BasicType.REAL)),
+    (SymbolAttributes(BasicType.DEFERRED), SymbolAttributes(BasicType.DEFERRED, intent='in'),
+     SymbolAttributes(BasicType.DEFERRED, intent='in')),  # Special case: Add attribute only
+    # Invalidate type by setting to DEFERRED
+    (SymbolAttributes(BasicType.INTEGER), SymbolAttributes(BasicType.DEFERRED),
+     SymbolAttributes(BasicType.DEFERRED)),
+    (SymbolAttributes(BasicType.REAL), SymbolAttributes(BasicType.DEFERRED),
+     SymbolAttributes(BasicType.DEFERRED)),
+    (SymbolAttributes(BasicType.DEFERRED, intent='in'), SymbolAttributes(BasicType.DEFERRED),
+     SymbolAttributes(BasicType.DEFERRED)),
+])
+def test_variable_clone_type(initype, newtype, reftype):
+    """
+    Test type updates are handled as expected and types are never ``None``.
+    """
+    scope = Scope()
+    var = symbols.Variable(name='var', scope=scope, type=initype)
+    assert 'var' in scope.symbol_attrs
+    new = var.clone(type=newtype)
```

Comment:
This just needs a `# pylint: disable=no-member` annotation.
(We really ought to convert the `Variable` factory class to a plain function)